File: erpcloud_itagksa/itag_manufacturing/work_order/work_order.py
# -*- coding: utf-8 -*-
# Copyright (c) 2026, Globcom Manufacturing and contributors
# For license information, please see license.txt

"""
Work Order Controller

Auto-fetches serial numbers from Stock Entries when Work Order
is created from Sales Order.

Validates that GRN exists for subcontracted Sales Orders before
allowing Work Order creation.
"""

import frappe
from frappe import _
from erpcloud_itagksa.itag_manufacturing.utils.work_order_serial_fetching import (
    fetch_serials_from_stock_entries
)
from erpcloud_itagksa.itag_manufacturing.sales_order.sales_order import validate_subcontract_grn_exists


def validate(doc, method=None):
    """Validate Work Order before save.

    Called via doc_events hook in hooks.py.

    Args:
        doc: Work Order document
        method: Event method name (not used)

    Validations:
    - If linked Sales Order is subcontracted, GRN must exist
    - If linked to a Stock Entry, populate subcontracted flag and serial fields
    """
    # Validate GRN exists for subcontracted Sales Orders
    if doc.sales_order:
        validate_subcontract_grn_exists(doc.sales_order)

    # Populate fields from linked Stock Entry (SE → WO flow)
    if doc.custom_stock_entry:
        populate_fields_from_stock_entry(doc)


def after_insert(doc, method=None):
    """Auto-fetch serial numbers after Work Order creation.

    Called via doc_events hook in hooks.py.

    Args:
        doc: Work Order document
        method: Event method name (not used)
    """
    # Fetch serials from Stock Entries linked via Sales Order (existing flow)
    fetch_serials_from_stock_entries(doc)


def populate_fields_from_stock_entry(work_order):
    """Populate Work Order fields from directly linked Stock Entry.

    Called during validate when custom_stock_entry is set. Handles:
    1. Sets custom_subcontracted_job to match the source Stock Entry.
    2. For each required_item whose serial fields are empty, copies
       serial_no, heat_no, heat_lot_no, drawing_no, and part_no from
       the matching Stock Entry Detail row (matched by item_code).

    Only fills empty fields — does not overwrite values the user has set.

    Args:
        work_order: Work Order document (in-memory, modifications persist on save)
    """
    stock_entry = frappe.get_doc("Stock Entry", work_order.custom_stock_entry)

    # Mirror the subcontracted flag from the source Stock Entry
    work_order.custom_subcontracted_job = stock_entry.custom_subcontracted_job

    if not work_order.required_items:
        return

    # Build a lookup of SE items by item_code (first match per code)
    se_items_by_code = {}
    for se_item in stock_entry.items:
        if se_item.item_code not in se_items_by_code:
            se_items_by_code[se_item.item_code] = se_item

    if not se_items_by_code:
        return

    for wo_item in work_order.required_items:
        se_item = se_items_by_code.get(wo_item.item_code)
        if not se_item:
            continue

        # Only populate serial fields that are currently empty
        if wo_item.custom_serial_no:
            continue

        wo_item.custom_serial_no = se_item.get("serial_no") or ""
        wo_item.custom_heat_no = se_item.get("custom_heat_no") or ""
        wo_item.custom_heat_lot_no = se_item.get("custom_heat_lot_no") or ""
        wo_item.custom_drawing_no = se_item.get("custom_drawing_no") or ""
        wo_item.custom_part_no = se_item.get("custom_part_no") or ""

File: erpcloud_itagksa/itag_manufacturing/work_order/work_order.js
// Copyright (c) 2026, Globcom Manufacturing and contributors
// For license information, please see license.txt

/**
 * Work Order Client Script
 *
 * Features:
 * - Auto-fetches heat_no, drawing_no, and part_no from Serial No master
 * - Populates parent custom_serial_no field with options from required_items
 */

frappe.ui.form.on('Work Order', {
    refresh: function(frm) {
        // Populate serial number options from required_items
        populate_serial_no_options(frm);
    },

    required_items_add: function(frm) {
        // Refresh serial number options when items are added
        populate_serial_no_options(frm);
    },

    required_items_remove: function(frm) {
        // Refresh serial number options when items are removed
        populate_serial_no_options(frm);
    }
});

frappe.ui.form.on('Work Order Item', {
    custom_serial_no: function(frm, cdt, cdn) {
        // Get the current row
        let row = locals[cdt][cdn];

        // Only auto-fetch if custom_serial_no is entered
        if (row.custom_serial_no) {
            auto_fetch_serial_attributes(frm, row);
        } else {
            // Clear custom fields if custom_serial_no is cleared
            frappe.model.set_value(cdt, cdn, 'custom_heat_no', '');
            frappe.model.set_value(cdt, cdn, 'custom_drawing_no', '');
            frappe.model.set_value(cdt, cdn, 'custom_part_no', '');
        }

        // Update parent field options when child serial number changes
        populate_serial_no_options(frm);
    }
});


/**
 * Populate parent custom_serial_no field with options from required_items child table
 */
function populate_serial_no_options(frm) {
    if (!frm.doc.required_items || frm.doc.required_items.length === 0) {
        // No items - clear options
        frm.set_df_property('custom_serial_no', 'options', '');
        return;
    }

    // Collect unique serial numbers from required_items
    let serial_numbers = new Set();

    frm.doc.required_items.forEach(function(item) {
        if (item.custom_serial_no) {
            // Split by newline or comma to handle multiple serials
            let serials = item.custom_serial_no.split(/[\n,]+/).map(s => s.trim()).filter(s => s);
            serials.forEach(serial => serial_numbers.add(serial));
        }
    });

    // Convert Set to sorted array
    let options_array = Array.from(serial_numbers).sort();

    // Set as options for the parent field (newline-separated)
    let options_string = options_array.join('\n');
    frm.set_df_property('custom_serial_no', 'options', options_string);

    // AUTO-POPULATE: If only 1 serial number, auto-select it (UX improvement)
    // Use direct assignment instead of frm.set_value() to avoid marking the
    // form dirty on post-save refresh. The value is displayed immediately and
    // persisted on the user's next save.
    if (options_array.length === 1 && !frm.doc.custom_serial_no) {
        frm.doc.custom_serial_no = options_array[0];
        frm.refresh_field('custom_serial_no');
        frappe.show_alert({
            message: __('Serial number auto-selected: {0}', [options_array[0]]),
            indicator: 'blue'
        }, 3);
    }

    // Refresh the field to show updated options
    frm.refresh_field('custom_serial_no');
}


/**
 * Auto-fetch serial attributes from Serial No master
 */
function auto_fetch_serial_attributes(frm, row) {
    // Show loading indicator
    frappe.dom.freeze(__('Fetching serial attributes...'));

    // Call server-side method to fetch attributes
    frappe.call({
        method: 'erpcloud_itagksa.itag_manufacturing.utils.serial_field_utils.fetch_serial_attributes_from_master',
        args: {
            serial_nos_string: row.custom_serial_no
        },
        callback: function(r) {
            frappe.dom.unfreeze();

            if (r.message) {
                let data = r.message;

                // Convert arrays to newline-separated strings
                let heat_nos_string = data.heat_nos.join('\n');
                let drawing_nos_string = data.drawing_nos.join('\n');
                let part_nos_string = data.part_nos.join('\n');

                // Update the row fields
                frappe.model.set_value(row.doctype, row.name, 'custom_heat_no', heat_nos_string);
                frappe.model.set_value(row.doctype, row.name, 'custom_drawing_no', drawing_nos_string);
                frappe.model.set_value(row.doctype, row.name, 'custom_part_no', part_nos_string);

                // Show success message if attributes were found
                if (heat_nos_string || drawing_nos_string || part_nos_string) {
                    frappe.show_alert({
                        message: __('Serial attributes auto-populated'),
                        indicator: 'green'
                    }, 3);
                } else {
                    // Show info message if no attributes found
                    frappe.show_alert({
                        message: __('No attributes found for these serial numbers'),
                        indicator: 'orange'
                    }, 3);
                }
            }
        },
        error: function(r) {
            frappe.dom.unfreeze();
            frappe.msgprint({
                title: __('Error'),
                message: __('Failed to fetch serial attributes. Please check the serial numbers.'),
                indicator: 'red'
            });
        }
    });
}
